platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/BestEffortPreconsensusEventFileCopy.java

```diff
@@ -0,0 +1,273 @@
+
+    /**
+     * The number of times to attempt to copy the last PCES file. Access to this file is not really coordinated between
+     * this logic and the code responsible for managing PCES file lifecycle, and so there is a small chance that the
+     * file moves when we attempt to make a copy. However, this probability is fairly small, and it is very unlikely
+     * that we will be unable to snatch a copy in time with a few retries.
+     */
+    private static final int COPY_PCES_MAX_RETRIES = 10;
+
+    private BestEffortPreconsensusEventFileCopy() {}
+
+    /**
+     * Copy preconsensus event files into the signed state directory. Copying these files is not thread safe and may
+     * fail as a result. This method retries several times if a failure is encountered. Success is not guaranteed, but
+     * success or failure is atomic and will not throw an exception.
+     *
+     * @param platformContext             the platform context
+     * @param selfId                      the id of this node
+     * @param destinationDirectory        the directory where the state is being written
+     * @param minimumGenerationNonAncient the minimum generation of events that are not ancient, with respect to the
+     *                                    state that is being written
+     */
+    public static void copyPreconsensusEventStreamFilesRetryOnFailure(
+            @NonNull final PlatformContext platformContext,
+            @NonNull final NodeId selfId,
+            @NonNull final Path destinationDirectory,
+            final long minimumGenerationNonAncient) {
+
+        final boolean copyPreconsensusStream = platformContext
+                .getConfiguration()
+                .getConfigData(PreconsensusEventStreamConfig.class)
+                .copyRecentStreamToStateSnapshots();
+        if (!copyPreconsensusStream) {
+            // PCES copying is disabled
+            return;
+        }
+
+        final Path pcesDestination =
+                destinationDirectory.resolve("preconsensus-events").resolve(Long.toString(selfId.id()));
+
+        int triesRemaining = COPY_PCES_MAX_RETRIES;
+        while (triesRemaining > 0) {
+            triesRemaining--;
+            try {
+                executeAndRename(
+                        pcesDestination,
+                        temporaryDirectory -> copyPreconsensusEventStreamFiles(
+                                platformContext, selfId, temporaryDirectory, minimumGenerationNonAncient));
+
+                return;
+            } catch (final IOException e) {
+                if (triesRemaining > 0) {
+                    logger.warn(STATE_TO_DISK.getMarker(), "Unable to copy PCES files. Retrying.");
+                } else {
+                    logger.error(
+                            EXCEPTION.getMarker(),
+                            "Unable to copy the last PCES file after {} retries. "
+                                    + "PCES files will not be written into the state.",
```

Comment:
Not required. It would be nice to include in the log message the saved state directory that will not have the PCES files. That way it is very easy to associate this log message with the particular saved state that cannot be used.

Reply:
Added round number to the log.